sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveInspectors.scala

```diff
@@ -213,8 +215,16 @@ private[hive] trait HiveInspectors {
 
     case c: Class[_] if c.isArray => ArrayType(javaClassToDataType(c.getComponentType))
 
+    // list type
+    case c: Class[_] if c == classOf[java.util.List[java.lang.Object]] =>
+      logWarning("Failed to catch a correct component type in List<> because of type erasure," +
+        " so you need to handle it correctly by yourself")
+      ArrayType(ErasedType)
+
     // Hive seems to return this for struct types?
     case c: Class[_] if c == classOf[java.lang.Object] => NullType
+
+    case c => throw new HiveDataTypeException("Unknown java type: " + c)
```

Comment:
s"Unsupported java type $c" seems to be better in this error message because this method is not only designed for UDF.